Add get all peers to GRPC

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This provides a call to grpc to get the peers a node is connected to. 
Adds an integration test for peer syncing
Adds an allow_test_network flag config option to allow nodes to use addresses such as 127.0.0.1 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This will allow integration tests as well as insights into how nodes are connected. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -61,6 +61,8 @@ service BaseNode {
     rpc FetchMatchingUtxos(FetchMatchingUtxosRequest) returns (stream FetchMatchingUtxosResponse);
 
+    // get all peers from the base node
+    rpc GetPeers(Empty) returns (GetPeersResponse);
```
Add a GetPeersRequest, even if it's empty. This way we can add arguments without breaking the signature.  Some arguments might be filtering by online, basenode/wallet, supported features. etc.

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -223,3 +225,8 @@ message FetchMatchingUtxosRequest {
+
+// This is the request type of the get all peers rpc call
+message GetPeersResponse{
+    repeated Peer peers = 1;
```
Maybe this should be streamed?

SWvheerden replied:
How big can this be?

sdbondi replied:
That is a good question, one that can be answered by writing a little test that serialises a peer and outputs the byte size, then expect 10's of thousands of peers (monero's grey list is 5000 big, but we also have client nodes (e.g. mobile wallets)). I recommend streaming because:
1. we can iterate through the peers instead of loading all of them in memory at once (correction: peer manager doesnt support an iterator but could/should in future)
1. faster time to start receiving and processing peers
1. we dont have to worry about the maximum message size being reached (not sure what that is on tonic grpc)
1. streaming is supported and not very hard to do

Of course, this won't be an issue right now, but as the peer list grows this will slow down

stringhandler commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,43 @@ message ConsensusConstants {
+    string added_at= 13;
+    /// User agent advertised by the peer
+    string user_agent = 14; 
+    map<uint32,bytes> metadata = 15;
```
Metadata can probably be excluded

stringhandler replied:
Otherwise this peer is technically vouching for the other peer

stringhandler commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,43 @@ message ConsensusConstants {
+    /// used as information for more efficient protocol negotiation.
+    repeated bytes supported_protocols = 12;
+    /// Timestamp of when the peer was added to this nodes peer list
+    string added_at= 13;
```
Remove

stringhandler commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,43 @@ message ConsensusConstants {
+    /// Connection statics for the peer
+    string last_connected_at = 10;
+    /// Represents the last connection attempt
+    string last_connection_attempt=11;
```
remove

stringhandler commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,43 @@ message ConsensusConstants {
+    /// Features supported by the peer
+    string features = 9;
+    /// Connection statics for the peer
+    string last_connected_at = 10;
```
String seems like the wrong data type for times

SWvheerden replied:
Its a date. Could perhaps convert this to a i64 so it represents a unix timestamp?

stringhandler replied:
protobuf `Timestamp` maybe?

stringhandler commented on base_layer/core/tests/wallet.rs:
```diff
@@ -108,264 +108,264 @@ fn create_peer(public_key: CommsPublicKey, net_address: Multiaddr) -> Peer {
-    let alice_temp_dir = tempdir().unwrap();
-    let bob_temp_dir = tempdir().unwrap();
-    let factories = CryptoFactories::default();
+// #[test]
```
What's the reason for this?

SWvheerden replied:
Removed as it was broken at the time of writing. The merge conflict should fix this.

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    /// Public key of the peer
+    bytes public_key =1;
+    /// NodeId of the peer
+    string node_id =2;
```
```suggestion
    bytes node_id =2;
```

sdbondi replied:
You could also derive this from the public key to save a few bytes `NodeId::from_public_key`

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+
+impl From<Peer> for grpc::Peer {
+    fn from(peer: Peer) -> Self {
+        let public_key = Vec::from(peer.public_key.as_bytes());
+        let node_id = peer.node_id.short_str();
```
Use the ByteArray impl
```suggestion
        let public_key = peer.public_key.to_vec();
        let node_id = peer.node_id.to_vec();
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    /// Peer's addresses
+    repeated Address addresses = 3;
+    /// Last connection attempt to peer
+    string last_connection = 4;
```
At the top of the proto file
```protobuf
import "google/protobuf/timestamp.proto";
```

```suggestion
    google.protobuf.Timestamp  last_connection = 4;
```

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+    fn from(peer: Peer) -> Self {
+        let public_key = Vec::from(peer.public_key.as_bytes());
+        let node_id = peer.node_id.short_str();
+        let mut addresses: Vec<grpc::Address> = Vec::new();
+        let last_connection = match peer.addresses.last_seen() {
+            Some(v) => v.to_string(),
+            None => "".to_string(),
+        };
+        for address in peer.addresses.addresses {
+            addresses.push(address.clone().into())
+        }
```
Something like this should work, you'll need the datetime_to_timestamp function available in this crate

```suggestion
        let addresses = peer.addresses.addresses.into_iter().map(Into::into).collect();     
        let last_connection = peer.addresses.last_seen().map(datetime_to_timestamp);             
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    /// Last connection attempt to peer
+    string last_connection = 4;
+    /// Flags for the peer.
+    string flags = 5;
```
```suggestion
    uint64 flags = 5;
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    string last_connection = 4;
+    /// Flags for the peer.
+    string flags = 5;
+    string banned_until= 6;
```
```suggestion
    google.protobuf.Timestamp banned_until= 6;
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    string flags = 5;
+    string banned_until= 6;
+    string banned_reason= 7;
+    string offline_at = 8;
```
```suggestion
    google.protobuf.Timestamp offline_at = 8;
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    string banned_reason= 7;
+    string offline_at = 8;
+    /// Features supported by the peer
+    string features = 9;
```
```suggestion
    uint64 features = 9;
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    /// Features supported by the peer
+    string features = 9;
+    /// Connection statics for the peer
+    string last_connected_at = 10;    /// Protocols supported by the peer. This should not be considered a definitive list of supported protocols and is
```
```suggestion
    // Protocols supported by the peer. This should not be considered a definitive list of supported protocols
    // and is used as information for more efficient protocol negotiation.
    google.protobuf.Timestamp last_connected_at = 10;   
```

sdbondi commented on applications/tari_app_grpc/proto/types.proto:
```diff
@@ -228,3 +228,37 @@ message ConsensusConstants {
+    uint64 avg_latency = 5;
+}
+
+message Peer{
```
Protobuf uses `//` for comments and the proto builder will make them into rust docstrings

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+        for address in peer.addresses.addresses {
+            addresses.push(address.clone().into())
+        }
+        let flags = if peer.flags == PeerFlags::empty() {
+            "".to_string()
+        } else {
+            format!("{:?}", peer.flags)
+        };
```
```suggestion
        let flags = peer.flags.bits();
```

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+        } else {
+            format!("{:?}", peer.flags)
+        };
+        let banned_until = match peer.banned_until {
+            Some(v) => v.to_string(),
+            None => "".to_string(),
+        };
```
```suggestion
        let banned_until = peer.banned_until.map(datetime_to_timestamp);
```

sdbondi replied:
Same for the other timestamps - the timestamp type is much smaller than a string repr of a timestamp

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+            Some(v) => v.to_string(),
+            None => "".to_string(),
+        };
+        let features = match peer.features {
+            PeerFeatures::COMMUNICATION_NODE => "BASE_NODE".to_string(),
+            PeerFeatures::COMMUNICATION_CLIENT => "WALLET".to_string(),
+            f => format!("{:?}", f),
+        };
```
The receiver of the peer data should make the base node/wallet determination - in future we could have validator node in the features (though likely it's easier to just have a PeerRole enum). 
 
```suggestion
        let features = match peer.features.bits();
          
```

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+            PeerFeatures::COMMUNICATION_CLIENT => "WALLET".to_string(),
+            f => format!("{:?}", f),
+        };
+        let last_connected_at = match peer.connection_stats.last_connected_at {
+            Some(v) => v.to_string(),
+            None => "".to_string(),
+        };
```
datetime_to_timestamp

sdbondi commented on applications/tari_app_grpc/src/conversions/peer.rs:
```diff
@@ -0,0 +1,106 @@
+            Some(v) => v.to_string(),
+            None => "".to_string(),
+        };
+        let mut supported_protocols: Vec<Vec<u8>> = Vec::new();
+        for proto in peer.supported_protocols {
+            supported_protocols.push(proto.clone().to_vec());
+        }
```
Using an iterator also preallocates the Vec once (as in `Vec::with_capacity`)

```suggestion
        let supported_protocols = peer.supported_protocols.into_iter().map(|p| p.to_vec()).collect();
```

sdbondi reviewed (commented):
Looks good - we should use the correct types for the proto messages

stringhandler commented on integration_tests/features/Sync.feature:
```diff
@@ -22,3 +22,11 @@ Feature: Block Sync
+  Scenario: When a new node joins the network, it should receive all peers. 
+  Given I have 10 seed nodes
+  And I have a base node NODE1 connected to all seed nodes
+  Then NODE1 should have 8 peers
```
Shouldn't this be 10?
```suggestion
  Then NODE1 should have 10 peers
```

stringhandler commented on integration_tests/features/Sync.feature:
```diff
@@ -22,3 +22,11 @@ Feature: Block Sync
+  And I have a base node NODE1 connected to all seed nodes
+  Then NODE1 should have 8 peers
+  Given I have a base node NODE2 connected to seed NODE1
+  Then NODE1 should have 8 peers
```
```suggestion
  Then NODE1 should have 11 peers
```

stringhandler commented on integration_tests/features/Sync.feature:
```diff
@@ -22,3 +22,11 @@ Feature: Block Sync
+  Then NODE1 should have 8 peers
+  Given I have a base node NODE2 connected to seed NODE1
+  Then NODE1 should have 8 peers
+  Then NODE2 should have 8 peers
```
```suggestion
  Then NODE2 should have 11 peers
```

stringhandler reviewed (dismissed):
Looks good, just need to fix the clippy error

SWvheerden committed:
Add proto files for getpeers

add integration test for getpeers
fix integration test
Add allow_test_address to config
